File: src/main/java/io/gittul/domain/comment/CommentService.java
package io.gittul.domain.comment;

import io.gittul.domain.comment.dto.CommentCreateRequest;
import io.gittul.domain.comment.dto.CommentResponse;
import io.gittul.domain.comment.entity.Comment;
import io.gittul.domain.post.PostRepository;
import io.gittul.domain.post.entity.Post;
import io.gittul.domain.user.entity.User;
import io.gittul.global.exception.CustomException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class CommentService {

    private final PostRepository postRepository;

    public CommentResponse createComment(CommentCreateRequest request,
                                         Long postId,
                                         User user) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new CustomException(HttpStatus.NOT_FOUND, "게시글을 찾을 수 없습니다."));


        Comment comment = Comment.of(request.content(), request.image(), user, post);
        post.getComments().add(comment);
        postRepository.save(post);

        return CommentResponse.ofAndTo(comment, user);
    }

    public void deleteComment(User user, Long postId, Long id) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new CustomException(HttpStatus.NOT_FOUND, "게시글을 찾을 수 없습니다."));

        Comment requestingComment = post.getComments().stream()
                .filter(comment -> comment.getCommentId().equals(id))
                .findAny()
                .orElseThrow(() -> new CustomException(HttpStatus.NOT_FOUND, "댓글을 찾을 수 없습니다."));

        if (!requestingComment.getUser().getUserId().equals(user.getUserId())) {
            throw new CustomException(HttpStatus.FORBIDDEN, "댓글을 삭제할 권한이 없습니다.");
        }

        post.getComments().remove(requestingComment);
        postRepository.save(post);
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/comment/CommentService.kt
package io.gittul.app.domain.comment

import io.gittul.app.domain.comment.dto.CommentCreateRequest
import io.gittul.app.domain.thread.repository.ThreadRepository
import io.gittul.core.domain.comment.entity.Comment
import io.gittul.core.domain.user.entity.User
import io.gittul.core.global.exception.CustomException
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class CommentService(
    private val threadRepository: ThreadRepository,
    private val commentRepository: CommentRepository
) {

    @Transactional
    fun createComment(
        request: CommentCreateRequest,
        threadId: Long,
        user: User?
    ): Comment {
        val thread = threadRepository.findById(threadId)
            .orElseThrow {
                CustomException(
                    HttpStatus.NOT_FOUND,
                    "게시글을 찾을 수 없습니다."
                )
            }


        val comment = Comment.of(request.content, request.image, user, thread)
        thread.comments.add(comment)
        threadRepository.saveAndFlush(thread)

        return comment
    }

    fun deleteComment(user: User, id: Long) {
        val requestingComment = commentRepository.findById(id).orElseThrow {
            CustomException(HttpStatus.NOT_FOUND, "댓글을 찾을 수 없습니다.")
        }

        val thread = requestingComment.thread

        if (requestingComment.user.userId != user.userId) {
            throw CustomException(HttpStatus.FORBIDDEN, "댓글을 삭제할 권한이 없습니다.")
        }

        thread.comments.remove(requestingComment)
        threadRepository.save(thread)
    }
}
